homeassistant/components/sonos/__init__.py

```diff
@@ -47,9 +85,116 @@ async def async_setup(hass, config):
     """Set up Sonos from a config entry."""
+    pysonos.config.EVENTS_MODULE = events_asyncio
+
+    if DATA_SONOS not in hass.data:
+        hass.data[DATA_SONOS] = SonosData()
+
+    config = hass.data[DOMAIN].get("media_player", {})
+    _LOGGER.debug("Reached async_setup_entry, config=%s", config)
+
+    advertise_addr = config.get(CONF_ADVERTISE_ADDR)
+    if advertise_addr:
+        pysonos.config.EVENT_ADVERTISE_IP = advertise_addr
+
+    def _stop_discovery(event: Event) -> None:
+        data = hass.data[DATA_SONOS]
+        if data.discovery_thread:
+            data.discovery_thread.stop()
+            data.discovery_thread = None
+        if data.hosts_heartbeat:
+            data.hosts_heartbeat()
+            data.hosts_heartbeat = None
+
+    def _discovery(now: datetime.datetime | None = None) -> None:
+        """Discover players from network or configuration."""
+        hosts = config.get(CONF_HOSTS)
+
+        def _discovered_player(soco: SoCo) -> None:
+            """Handle a (re)discovered player."""
+            try:
+                _LOGGER.debug("Reached _discovered_player, soco=%s", soco)
+
+                data = hass.data[DATA_SONOS]
+
+                if soco.uid not in data.discovered:
+                    _LOGGER.debug("Adding new entity")
+                    data.discovered[soco.uid] = soco
+                    # Set these early since device_info() needs them
+                    data.speaker_info[soco.uid] = soco.get_speaker_info(True)
+
+                    dispatcher_send(hass, SONOS_DISCOVERY_UPDATE, {"soco": soco})
+                else:
+                    entity = data.media_player_entities.get(soco.uid)
+                    if entity and (entity.soco == soco or not entity.available):
+                        hass.add_job(entity.async_seen(soco))
+
+                    entity = data.battery_entities.get(soco.uid)
+                    if entity and (entity.soco == soco or not entity.available):
+                        hass.add_job(entity.async_seen(soco))
+
+                # watch for this soco to become unavailable
+                seen_timers = data.seen_timers
+                if soco.uid in seen_timers:
+                    seen_timers[soco.uid]()
+                seen_timers[soco.uid] = hass.helpers.event.async_call_later(
+                    SEEN_EXPIRE_TIME.seconds, lambda: _disappeared_player(soco)
+                )
+
+            except SoCoException as ex:
+                _LOGGER.debug("SoCoException, ex=%s", ex)
+
+        def _disappeared_player(soco: SoCo) -> None:
+            """Handle a player that has disappeared."""
+            data = hass.data[DATA_SONOS]
+            del data.seen_timers[soco.uid]
+            entity = data.media_player_entities.get(soco.uid)
+            if entity:
+                entity.async_unseen()
+            entity = data.battery_entities.get(soco.uid)
+            if entity:
+                entity.async_unseen()
```

Comment:
We should use the dispatching instead of storing the entities in a shared container